plenoptic/synthesize/geodesic.py

```diff
@@ -1,155 +1,304 @@
-        self.optimizer.step()
+        Parameters
+        ----------
+        max_iter: int, optional
+            maximum number of steps taken by the optimization a
 
-        self.t0 = self.t1
-        self.t1 = time.time()
-        pretty_print(i, max_iter, (self.t1 - self.t0), loss.item(), g.norm().item())
+        learning_rate: float, optional
+            controls the step sizes of the search algorithm
 
-        return loss
+        optimizer: str ('adam' or 'sgd') or torch.optim.Optimizer, optional
+            algorithm that will perform the search
 
-    def synthesize(self, max_iter=20, learning_rate=.01, objective='multiscale', seed=0):
+        lmbda: float, optional
+            strength of the regularizer that enforces the image range in [0, 1]
 
-        torch.manual_seed(seed)
-        self.optimizer = optim.Adam([self.x], lr=learning_rate, amsgrad=True)
+        objective: str, optional
+            'default'
+            'multiscale'
 
-        self.loss = []
-        self.t1 = time.time()
+        seed: int
+            set the random number generator
+        """
+        self.lmbda = lmbda
 
-        for i in range(max_iter):
-            loss = self._optimizer_step(i, max_iter)
+        torch.manual_seed(seed)
+        if optimizer == 'adam':
+            self.optimizer = optim.Adam([self.x],
+                                        lr=learning_rate, amsgrad=True)
+        elif optimizer == 'sgd':
+            self.optimizer = optim.SGD([self.x],
+                                       lr=learning_rate, momentum=0.9)
+        elif isinstance(optimizer, optim.Optimizer):
+            self.optimizer = optimizer
+
+        pbar = tqdm(range(max_iter))
+        for i in pbar:
+            loss = self._optimizer_step(i, pbar)
+
+            # storing some information
             self.loss.append(loss.item())
-
-        self.gamma.copy_(self.img.data.squeeze()).add_(0.5).mul_(255)
-
-        return self.gamma
+            self.geodesic = torch.cat((self.xA, self.x.data, self.xB), 0)
+            # TODO flag to store progress or not
+            self.dist_from_line.append(distance_from_line(
+                                        self.geodesic))
+
+            if loss.item() < 1e-6:
+                raise Exception("""the geodesic matches the representation
+                                straight line up to floating point
+                                precision""")
+
+    def plot_loss(self):
+        plt.semilogy(self.loss)
+        plt.xlabel('iter step')
+        plt.ylabel('loss value')
+        plt.show()
+
+    def plot_distance_from_line(self, vid=None, iteration=None,
+                                figsize=(7, 5)):
+        """visual diagnostic of geodesic linearity in representation space.
+
+        Parameters
+        ----------
+        vid : torch.Tensor, optional
+            natural video that bridges the anchor points
+        iteration : int, optional
```

Comment:
I find `vid` confusing. It's the "target", right? The actual natural video we're hoping our model's geodesic looks like? Could you rename it and expand upon that somewhere (I think it's only referred to here and in the animation, right)?